### PR TITLE
Framework level revision on particles

### PR DESCRIPTION
**Pull request by Xiangyu-Hu**
The main changes are:
1. Only geometric particles is defined, which suggested all physical variables are moved from particles classes and defined  in the particle dynamics method when it will used.
2. The definition rule of the a physical variable is, a) if the variable will be modified by the method, a shared variable is defined, b) otherwise the variable is get from the variables assemble.
3. The  old function defineParticlesAndMeterial in SPHBody class is changed to defineMaterial. Since only geometric particles are defined, the particle definition is done in the function generateParticles as template parameter.
4. The reaction and diffusion material and corresponding particle dynamics are revised so that the one can define single diffusion material and directly use it. The class ReactionDiffusion now is only used for reaction and diffusion system in which multiple species are involved.
5.  In contact diffusion dynamics, inter-body transfer is defined and is first computed as heat or mass transfer between bodies. Then, this transfer is added to the total diffusion rate of the body. With this change, one can easily obtained the interface heat transfer.

### CODE REVIEW

**Comment by DrChiZhang** [comment]
When the name of a member template specialization appears after . or -> in a postfix-expression, or after nested-name-specifier in a qualified-id, and the postfix-expression or qualified-id explicitly depends on a template-parameter (14.6.2), the member template name must be prefixed by the keyword template. Otherwise the name is assumed to name a non-template.

you may use
object->template f<T>();

**Comment by DrChiZhang** [comment]
When access to variable, for example "pos", some cases, for example, through " basec_particles.pos", other cases for example " dynamics", through "getVariableByname", why not only use "getVariableByname" ?

File: src/shared/materials/elastic_solid.h [hunk 1]
```diff
@@ -102,7 +102,10 @@ class ElasticSolid : public Solid
     virtual Real VolumetricKirchhoff(Real J) = 0;
     /** Define the calculation of the stress matrix for postprocessing */
     virtual std::string getRelevantStressMeasureName() = 0;
-
+    /** Get  average velocity when interacting with fluid. */
+    virtual StdLargeVec<Vecd> *AverageVelocity(BaseParticles *base_particles) override;
+    /** Get  average acceleration when interacting with fluid. */
+    virtual StdLargeVec<Vecd> *AverageAcceleration(BaseParticles *base_particles) override;
```
[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply src/shared/materials/elastic_solid.h:108]: They are. As rigid body use velocity directly but for deformable solid, it is a new variable.

**Comment by Xiangyu-Hu** [comment]
> When the name of a member template specialization appears after . or -> in a postfix-expression, or after nested-name-specifier in a qualified-id, and the postfix-expression or qualified-id explicitly depends on a template-parameter (14.6.2), the member template name must be prefixed by the keyword template. Otherwise the name is assumed to name a non-template.
> 
> you may use
> 
> ` object->template foo<templatename>();`

Sometime, the gcc thinks that is not necessary. But Clang is strict.

**Comment by Xiangyu-Hu** [comment]
> When access to variable, for example "pos", some cases, for example, through " basec_particles.pos", other cases for example " dynamics", through "getVariableByname", why not only use "getVariableByname" ?

Could you specify the cases as you pointed out?

**Comment by Xiangyu-Hu** [comment]
> When access to variable, for example "pos", some cases, for example, through " basec_particles.pos", other cases for example " dynamics", through "getVariableByname", why not only use "getVariableByname" ?

@DrChiZhang I have cleared this issue. Could you have a check?

**Review by Lisu-lisa — APPROVED** [review]

**Pull request merged**